packages/flutter_tools/lib/src/migrate/migrate_utils.dart

```diff
@@ -0,0 +1,325 @@
+  static Future<DiffResult> diffFiles(File one, File two, Logger logger) async {
+    if (one.existsSync() && !two.existsSync()) {
+      return DiffResult.deletion();
+    }
+    if (!one.existsSync() && two.existsSync()) {
+      return DiffResult.addition();
+    }
+    final List<String> cmdArgs = <String>['diff', '--no-index', one.absolute.path, two.absolute.path];
+    final ProcessResult result = await Process.run('git', cmdArgs);
+
+    // diff exits with 1 if diffs are found.
+    checkForErrors(result, logger, allowedExitCodes: <int>[1], commandDescription: 'git ${cmdArgs.join(' ')}');
+    return DiffResult(result);
+  }
+
+  // Clones a copy of the flutter repo into the destination directory. Returns false if unsucessful.
+  static Future<bool> cloneFlutter(String revision, String destination, Logger logger) async {
+    // Use https url instead of ssh to avoid need to setup ssh on git.
+    List<String> cmdArgs = <String>['clone', '--single-branch', '--filter=blob:none', '--shallow-exclude=v1.0.0', 'https://github.com/flutter/flutter.git', destination];
+    ProcessResult result = await Process.run('git', cmdArgs);
+    checkForErrors(result, logger, commandDescription: 'git ${cmdArgs.join(' ')}');
+
+    cmdArgs.clear();
+    cmdArgs = <String>['reset', '--hard', revision];
+    result = await Process.run('git', cmdArgs, workingDirectory: destination);
+    if (!checkForErrors(result, logger, commandDescription: 'git ${cmdArgs.join(' ')}', exit: false)) {
+      return false;
+    }
+    return true;
+  }
+
+  /// Calls `flutter create` as a re-entrant command.
+  static Future<String> createFromTemplates(String flutterBinPath, {
+    required String name,
+    bool legacyNameParameter = false,
+    required String androidLanguage,
+    required String iosLanguage,
+    required String outputDirectory,
+    String? createVersion,
+    List<String> platforms = const <String>[],
+    required Logger logger,
+    required FileSystem fileSystem,
+  }) async {
+    final List<String> cmdArgs = <String>['create'];
+    if (!legacyNameParameter) {
+      cmdArgs.add('--project-name=$name');
+    }
+    cmdArgs.add('--android-language=$androidLanguage');
+    cmdArgs.add('--ios-language=$iosLanguage');
+    if (platforms.isNotEmpty) {
+      String platformsArg = '--platforms=';
+      for (int i = 0; i < platforms.length; i++) {
+        if (i > 0) {
+          platformsArg += ',';
+        }
+        platformsArg += platforms[i];
+      }
+      cmdArgs.add(platformsArg);
+    }
+    cmdArgs.add('--no-pub');
+    if (legacyNameParameter) {
+      cmdArgs.add(name);
+    } else {
+      cmdArgs.add(outputDirectory);
+    }
+    final ProcessResult result = await Process.run('$flutterBinPath/flutter', cmdArgs, workingDirectory: outputDirectory);
+    final String error = result.stderr as String;
+
+    // Catch errors due to parameters not existing.
+
+    // Old versions of the tool does not include the platforms option.
```

Comment:
The flutter version tag is not always reliable in representing what parameters are available as it is not fine enough, I also tried using the hash to see if the boundaries between commits that add parameters can reliably obtained, but depending on the branch, this method was also not reliable due to various branches and releases.

This technique is the most robust way to approach it. Since we are effectively limiting the backwards compatibility to version v1.0.0, this should be able to self-pare its way down to the minimal runnable command. I can add a counter also to restrict this from infinite looping if you would like.

Reply:
What do you mean that it is not fine enough? Can't we just determine the first tagged version that had a specific change? GitHub will even do this for you if you look at a specific commit in the web UI:
![The bold tag on the left is the latest release with this commit, the tag on the right is the earliest tag with this commit, which we want](https://user-images.githubusercontent.com/7856010/164546354-c5eb90f0-5f21-449e-b649-66f6c4de96ce.png).

Reply:
It is not uncommon for projects to be created off of a custom branch with a commit hash that is not on the master branch. This is especially true for our first party sample projects which were largely created with squashed local commits.

In these cases, we can't really say what version of flutter it was created with in reference to the tags and if we get this wrong, the command will fail and we would have to fall back to this system anyways.

Reply:
Why only limit to 1.0? Where does the `--platforms` flag first appear? If it helps to delete this logic to instead require creation after 2.0 for migration to work, I think that would be totally reasonable.